### PR TITLE
executor: move into `tokio` crate

### PR DESCRIPTION
**Pull request by carllerche**
A step towards collapsing Tokio sub crates into a single `tokio`
crate (#1318).

The executor implementation is now provided by the main `tokio` crate.
Functionality can be opted out of by using the various net related
feature flags.

### CODE REVIEW

File: tokio/src/fs/mod.rs [hunk 1]
```diff
@@ -18,12 +18,10 @@
 //! type. Adaptions also extend to traits like `std::io::Read` where methods
 //! return `std::io::Result`.  Be warned that these adapted methods may return
 //! `std::io::ErrorKind::WouldBlock` if a *worker* thread can not be converted
-//! to a *backup* thread immediately. See [tokio-executor] for more details
-//! of the threading model and [`blocking`].
+//! to a *backup* thread immediately.
 //!
-//! [`blocking`]: https://docs.rs/tokio-executor/0.2.0-alpha.2/tokio_executor/threadpool/fn.blocking.html
+//! [`blocking`]: #
```
Comment (LucioFranco) [inline tokio/src/fs/mod.rs]: missing link?

File: tokio/src/executor/current_thread/scheduler.rs [hunk 2]
```diff
@@ -1,5 +1,5 @@
-use super::{Borrow, BorrowSpawner};
-use crate::park::Unpark;
+use crate::executor::current_thread::{Borrow, BorrowSpawner};
```
Comment (jonhoo) [inline tokio/src/executor/current_thread/scheduler.rs]: Why change this away from `super`?

File: tokio/src/executor/global.rs [hunk 3]
```diff
@@ -15,7 +15,7 @@ use std::pin::Pin;
 /// executor (usually itself) that is used to spawn new tasks.
 ///
 /// The current `DefaultExecutor` reference is tracked using a thread-local
-/// variable and is set using `tokio_executor::with_default`
+/// variable and is set using `crate::executor::with_default`
```
Comment (jonhoo) [inline tokio/src/executor/global.rs]: This looks wrong. Looks like a str-replace error, so probably happened elsewhere too.

File: tokio/src/executor/current_thread/scheduler.rs (hunk 2 above)
Comment (carllerche) [reply tokio/src/executor/current_thread/scheduler.rs]: I like being able to see the full path when reading code. I find it easier than doing the relative calculation based off of where I am at.

Also, always using absolute paths removes the decision of having to pick relative vs. crate absolute. Less decisions is better.

File: tokio/src/fs/mod.rs (hunk 1 above)
Comment (carllerche) [reply tokio/src/fs/mod.rs]: There is nothing to link to until #1681 lands. I opted to leave the comment vs. deleting it.

Comment (jonhoo) [inline tokio/src/fs/mod.rs]: Broken link.

File: tokio/src/lib.rs [hunk 4]
```diff
@@ -100,9 +105,11 @@ pub mod net;
 pub mod prelude;
 
 #[cfg(feature = "process")]
+#[cfg(not(loom))]
```
Comment (jonhoo) [inline tokio/src/lib.rs]: Why? And should this use `cfg(all())` ?

File: tokio/src/fs/mod.rs (hunk 1 above)
Comment (carllerche) [reply tokio/src/fs/mod.rs]: There is nothing to link to until #1681 lands. I opted to leave the comment vs. deleting it.

File: tokio/src/runtime/mod.rs [hunk 5]
```diff
@@ -19,7 +19,7 @@
 //! Creating a [`Runtime`] does the following:
 //!
 //! * Spawn a background thread running a [`Reactor`] instance.
-//! * Start a [`ThreadPool`] for executing futures.
+//! * Start a `ThreadPool` for executing futures.
```
Comment (jonhoo) [inline tokio/src/runtime/mod.rs]: Why the removal of this link? Also, can't we just rely on auto-doc-links, and not give the absolute URL?

File: tokio/tests/runtime_current_thread.rs [hunk 6]
```diff
@@ -1,5 +1,4 @@
 #![warn(rust_2018_idioms)]
-#![cfg(feature = "default")]
```
Comment (jonhoo) [inline tokio/tests/runtime_current_thread.rs]: Why is this removed?

File: tokio/Cargo.toml [hunk 7]
```diff
@@ -86,6 +87,7 @@ futures-util-preview = { version = "=0.3.0-alpha.19", features = ["sink"] }
 
 # Everything else is optional...
 bytes = { version = "0.4", optional = true }
+crossbeam-channel = { version = "0.3.8", optional = true }
```
Comment (jonhoo) [inline tokio/Cargo.toml]: Not sure I follow why this was added?

File: tokio/src/lib.rs (hunk 4 above)
Comment (carllerche) [reply tokio/src/lib.rs]: Tests don't work when running loom tests. I opted to disable the entire mod to speed up compilation time.

I am unsure if `all` is easier to read vs. two lines. I can switch though.

File: tokio/src/runtime/mod.rs (hunk 5 above)
Comment (carllerche) [reply tokio/src/runtime/mod.rs]: `ThreadPool` is going to go away in a follow up PR.

File: tokio/src/fs/mod.rs (hunk 1 above)
Comment (jonhoo) [reply tokio/src/fs/mod.rs]: I would just not _manually_ give the link, and have the rust auto-doc stuff take care of it. It will leave a warning if the symbol doesn't resolve. If it's left like this, we _will_ forget to change it.

File: tokio/tests/runtime_current_thread.rs (hunk 6 above)
Comment (carllerche) [reply tokio/tests/runtime_current_thread.rs]: We need better ways to organize tests & feature flags. I opted to punt to follow up work as the entire crate needs it.

File: tokio/Cargo.toml (hunk 7 above)
Comment (carllerche) [reply tokio/Cargo.toml]: current-thread needs it IIRC.

File: tokio/src/lib.rs (hunk 4 above)
Comment (jonhoo) [reply tokio/src/lib.rs]: We use `all` elsewhere, so we should probably stick to one or the other. I don't care which one.

I think I'd prefer to just disable the loom _tests_ for that module if possible, but don't consider this a blocker.

File: tokio/Cargo.toml (hunk 7 above)
Comment (jonhoo) [reply tokio/Cargo.toml]: Oh, I guess this was in `tokio-executor`'s `Cargo.toml` before then?

File: tokio/src/runtime/mod.rs (hunk 5 above)
Comment (hawkw) [reply tokio/src/runtime/mod.rs]: @carllerche if the `ThreadPool` type is going away, why not change this to 
```suggestion
//! * Start a thread pool for executing futures.
```

**Comment by hawkw** [comment]
Looks like this needs to be rebased, since #1625 made a bunch of conflicting changes.

File: tokio/Cargo.toml (hunk 7 above)
Comment (carllerche) [reply tokio/Cargo.toml]: yes

File: tokio/src/fs/mod.rs (hunk 1 above)
Comment (carllerche) [reply tokio/src/fs/mod.rs]: Done.

File: tokio/src/lib.rs (hunk 4 above)
Comment (carllerche) [reply tokio/src/lib.rs]: Done

**Comment by najamelan** [comment]
I was using the tokio-executor crate directly, to avoid bloating dependencies. It looks like currently the only way to get the functionality back is to enable the "rt-full" feature. 

There are many reasons not to want to pull all of this in. Maybe a library just wants to implement traits `Executor` and `TypedExecutor` without even pulling in an impl, maybe one writes an application which does FS access, but no networking, maybe one wants to use the Actor model as a software design model and do async even without IO, etc...

Would it be possible to design the features so as to be able to enable things bottom up, eg. not pulling anything not needed at each step? 

Slightly off topic: Other recent changes have made tokio harder to use flexibly. eg. Networking types will no longer work when not within a runtime (before it was possible to use the networking primitives and spawn their futures on a non-tokio executor). 

It's unfortunate that the reasoning is package-deal like. I am very reluctant to use frameworks like this, even though I feel tokio has an excellent executor implementation. It's all good when writing a small app, but when writing libraries that shall be used by others, there's no way to justify forcing the package deal on downstream users.

**Comment by taiki-e** [comment]
@najamelan FYI: Once #1695 is merged, it should be able to use traits without depending on the runtime: https://github.com/tokio-rs/tokio/pull/1695/files#diff-66fb4a030b180360b997fa738670656fR124-R125

**Comment by najamelan** [comment]
@taiki-e Thanks, so I see there will be `executor-core`. Will it be possible to use the 2 executor impls that existed in tokio-executor without pulling reactor/timer/network?

**Comment by carllerche** [comment]
@najamelan thanks for the heads up. The state if things is not final. The plan will be to enable the ability to get fine grained access to the various components using feature flags.

Collapsing crates is just the first step.

Could you open a new issue so we are sure to cover your use case?

**Comment by najamelan** [comment]
@carllerche Ok, thanks will open an issue.
